File: samestr/summarize/read_samestr_data.R
# libraries
suppressPackageStartupMessages(require(tidyverse))

# functions
read_distmat = function(directory, species, suffix) {
  fn = paste0(directory, '/', species, suffix)
  distmat = suppressWarnings(read.delim(fn, '\t', header=T, row.names=1, stringsAsFactors=F, check.names = F))
  return(distmat)
}

distmat_nonfinite_to_NA = function(distmat) {
  "
  Function takes distance matrix as data frame as input
  transforms non-finite values to NA, 
  returns distance matrix as data frame.
  "
  col_names <- colnames(distmat)
  row_names <- rownames(distmat)
  distmat <- do.call(data.frame, lapply(distmat, function(x) replace(x, !is.finite(x), NA)))
  colnames(distmat) <- col_names
  rownames(distmat) <- row_names
  return(distmat)
}

distmat_rm_upper_tri = function(distmat, rm_diag=FALSE) {
  "
  Function takes distance matrix as data frame as input
  and returns distance matrix as data frame without upper triangle.
  "  
  if (rm_diag == TRUE) {
    distmat[upper.tri(distmat, diag = TRUE)] <- NA
  } else {
    distmat[upper.tri(distmat, diag = FALSE)] <- NA
  }
  return(distmat)
}

distmat_to_long = function(distmat, value_name, rm_diag = TRUE) {
  "
  Function takes distance matrix as data frame as input
  and returns a data frame in long format,
  containing all values from the distance matrix:
  
  Requires: 
  At least lower triangle of distance matrix,
  modify_distmat script for distance matrix formatting.
  
  -- Output Format
  row col distance
  A B 2.5
  "
  distmat.long <- 
    distmat_rm_upper_tri(distmat, rm_diag = rm_diag) %>% 
    rownames_to_column('row') %>% 
    pivot_longer(names_to = 'col', 
                 values_to = value_name, 
                 cols = -matches('row'), 
                 values_drop_na = T)
  return(distmat.long)
}

get_merged_distmats = function(directory, species) {
  "
  Takes directory and species as input
  and returns a data.frame where file pairs for each species are merged.
  
  Requires:
  Input dir must contain samestr compare output files with
  extensions: .fraction.txt and .overlap.txt
  "
  distmat = read_distmat(directory, species, '.fraction.txt')
  overlap = read_distmat(directory, species, '.overlap.txt')
  
  # tag nan, inf as NA
  distmat = distmat_nonfinite_to_NA(distmat)
  overlap = distmat_nonfinite_to_NA(overlap)
  
  # wide to long
  distmat.long = distmat_to_long(distmat, value_name='similarity', rm_diag=T)
  overlap.long = distmat_to_long(overlap, value_name='overlap', rm_diag=T)
  
  if (nrow(distmat.long) & nrow(overlap.long)) {
    merged_distmat_overlap <- 
      
      full_join(distmat.long, overlap.long, 
                by = c('row','col')) %>% 
      mutate(species = species) %>% 
      filter(row != col)
  }
  return(merged_distmat_overlap)
}

# main
read_samestr_data <- function(sstr_dir) {
  "
  Takes samestr compare output directory as input and runs get_merged_distmats
  "
  distance_species_list <- gsub(list.files(path = sstr_dir, 
                                           pattern = '.fraction.txt', full.names = F), 
                                pattern = '.fraction.txt', replacement = '')
  overlap_species_list <- gsub(list.files(path = sstr_dir, 
                                          pattern = '.overlap.txt', full.names = F), 
                               pattern = '.overlap.txt', replacement = '')
  
  species_intersect = intersect(overlap_species_list, distance_species_list)
  print(paste0('Merging matrix for ', length(species_intersect), ' species.'))
  sstr_data = data.frame()
  for (species in species_intersect) {
    sstr_data = bind_rows(sstr_data, get_merged_distmats(directory=sstr_dir, species=species))
  }
  return(sstr_data)
}

File: samestr/summarize/read_samestr_data.py
from os import listdir
import pandas as pd
import numpy as np
import logging

LOG = logging.getLogger(__name__)


def read_distmat(directory, species, suffix):
    """
    Function reads distance matrix from file
    returns distance matrix as data frame.
    """
    fn = f"{directory}/{species}{suffix}"
    distmat = pd.read_csv(fn, sep="\t", index_col=0)
    return distmat


def distmat_nonfinite_to_NA(distmat):
    """
    Function takes distance matrix as data frame as input
    transforms non-finite values to NA, 
    returns distance matrix as data frame.
    """
    distmat = distmat.mask(~np.isfinite(distmat))
    return distmat


def distmat_rm_upper_tri(distmat, rm_diag=False):
    """
    Function takes distance matrix as data frame as input
    and returns distance matrix as data frame without upper triangle.
    It also removes diagonal if rm_diag is True.
    """
    # Remove upper triangle
    distmat = distmat.where(np.tril(np.ones(distmat.shape)).astype(bool))

    # Remove diagonal
    if rm_diag:
        distmat = distmat.where(~np.eye(distmat.shape[0], dtype=bool))
    return distmat


def distmat_to_long(distmat, value_name):
    """
    Function takes distance matrix as data frame as input
    and returns a data frame in long format,
    containing all values from the distance matrix.
    """
    # Convert to long format
    distmat = distmat.stack().reset_index()
    distmat.columns = ['row', 'col', value_name]
    return distmat


def get_merged_distmats(directory, clade):
    """
    Takes directory and clade as input
    and returns a data.frame where file pairs for each clade are merged.

    Requires:
    Input dir must contain samestr compare output files with
    extensions: .fraction.txt and .overlap.txt
    """
    # Read in the fraction and overlap files as dataframes
    distmat = read_distmat(directory, clade, '.fraction.txt')
    overlap = read_distmat(directory, clade, '.overlap.txt')

    # Convert non-finite values to NA
    distmat = distmat_nonfinite_to_NA(distmat)
    overlap = distmat_nonfinite_to_NA(overlap)

    # Convert wide to long format
    distmat_long = distmat_to_long(distmat, value_name='similarity')
    overlap_long = distmat_to_long(overlap, value_name='overlap')
    if distmat_long.shape[0] > 0 and overlap_long.shape[0] > 0:
        # Merge the two dataframes
        merged_distmat_overlap = pd.merge(distmat_long, overlap_long, on=[
                                          'row', 'col'], how='outer')

        # Add a column for clade
        merged_distmat_overlap['clade'] = clade

        # remove self-matches, one side of the matrix
        merged_distmat_overlap = merged_distmat_overlap[merged_distmat_overlap['row']
                                                        < merged_distmat_overlap['col']]
        return merged_distmat_overlap


def read_samestr_data(sstr_dir):
    """
    Takes samestr compare output directory as input and runs get_merged_distmats
    """
    distance_clade_list = [s.replace('.fraction.txt', '') for s in listdir(
        sstr_dir) if s.endswith('.fraction.txt')]
    overlap_clade_list = [s.replace('.overlap.txt', '') for s in listdir(
        sstr_dir) if s.endswith('.overlap.txt')]
    clade_intersect = set(overlap_clade_list).intersection(
        set(distance_clade_list))

    LOG.info(f'Merging matrix for {len(clade_intersect)} clades.')
    sstr_data = pd.DataFrame()
    for clade in clade_intersect:
        sstr_data = pd.concat(
            [sstr_data, get_merged_distmats(directory=sstr_dir, clade=clade)])

    return sstr_data


def analyze_strain_events(sstr_data, similarity_threshold=0.999, overlap_threshold=5000):
    """
    Takes samestr data as input and annotates it with analysis level, shared, and event.
    Also returns a data frame with the sum of strain co-occurrences
    across all clades for each file pair, including zero counts.
    """
    sstr_data['analysis_level'] = sstr_data.apply(
        lambda x: 'strain-level' if x['overlap'] > overlap_threshold else 'clade-level', axis=1)
    sstr_data['analyzed_strain'] = sstr_data['analysis_level'] == 'strain-level'
    sstr_data['shared_strain'] = (sstr_data['similarity'] >= similarity_threshold) & (
        sstr_data['analyzed_strain'] == True)
    sstr_data['event'] = sstr_data.apply(lambda x: 'shared_strain' if x['shared_strain']
                                         else 'other_strain' if x['overlap'] > overlap_threshold else 'same_clade', axis=1)
    sstr_data = sstr_data[['row', 'col', 'clade', 'similarity', 'overlap',
                           'analysis_level', 'analyzed_strain', 'shared_strain', 'event']]

    # get the sum of shared strains and analyzed strains for each file pair
    strain_cooccurrences = sstr_data.groupby(
        ['row', 'col'])[['shared_strain', 'analyzed_strain']].sum().reset_index()
    
    sstr_data = sstr_data.sort_values(["row", "col", "clade"])
        
    return sstr_data, strain_cooccurrences
